### PR TITLE
Reduce config requirements

### LINKED ISSUE #186: Reduce config requirements
Currently too many fields when creating the `App` are required, we should stop requiring them where it makes sense or merge fields in other cases.

### PR DESCRIPTION
**Pull request by terw-dan**
Fixes #186
Adds defaults for most of the config options for the app.

### CODE REVIEW

**Comment by jcomack** [comment]
CR done 👍

File: js/app.js [hunk 1]
```diff
@@ -97,9 +151,27 @@ function createDefaultSnippetPreview() {
...
-	this.config = this.extendConfig( args );
+	if ( !isObject( args ) ) {
+		args = {};
+	}
+	defaultsDeep( args, defaults );
+
+	this.config = args;
 	this.callbacks = this.config.callbacks;
 
+	if ( !isObject( this.callbacks.getData ) ) {
```
Comment (atimmer) [inline js/app.js]: I would do these checks before we're even assigning them to our `App` properties.

File: dist/yoast-seo.js [hunk 2]
```diff
@@ -1194,9 +1248,27 @@ function createDefaultSnippetPreview() {
...
+	}
+	defaultsDeep( args, defaults );
+
+	this.config = args;
 	this.callbacks = this.config.callbacks;
 
+	if ( !isObject( this.callbacks.getData ) ) {
+		throw new Error( "The app requires an object with a getdata callback." );
+	}
+	if ( !isObject( this.config.targets ) ) {
```
Comment (atimmer) [inline dist/yoast-seo.js]: Please put some newlines so it is more readable.

File: dist/yoast-seo.js [hunk 3]
```diff
@@ -1194,9 +1248,27 @@ function createDefaultSnippetPreview() {
...
-	this.config = this.extendConfig( args );
+	if ( !isObject( args ) ) {
+		args = {};
+	}
+	defaultsDeep( args, defaults );
+
+	this.config = args;
 	this.callbacks = this.config.callbacks;
 
+	if ( !isObject( this.callbacks.getData ) ) {
```
Comment (atimmer) [inline dist/yoast-seo.js]: Please put all the error throwing in a different private function for readability of the constructor.

File: js/app.js [hunk 4]
```diff
@@ -27,6 +81,26 @@ function createDefaultSnippetPreview() {
 	} );
 }
 
+function checkArguments( args ) {
```
Comment (atimmer) [inline js/app.js]: Please document this function. Also use the `@private` flag to indicate that this is a private function.

File: js/app.js [hunk 5]
```diff
@@ -97,7 +171,15 @@ function createDefaultSnippetPreview() {
...
 YoastSEO.App = function( args ) {
-	this.config = this.extendConfig( args );
+	if ( !isObject( args ) ) {
+		args = {};
+	}
+	defaultsDeep( args, defaults );
+
+	this.config = args;
+
+	checkArguments( this.config );
```
Comment (atimmer) [inline js/app.js]: We should move this above the `this.config = args` line. We should check inputs before we assign them to our instance properties.

**Comment by andizer** [comment]
CR done.

**Comment by andizer** [comment]
Please fix the conflicts.

**Comment by terw-dan** [comment]
Merge conflicts fixed

**Comment by boblinthorst** [comment]
acceptance seems ok :+1:

**Pull request merged**